extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcRedirectFilter.java

```diff
@@ -0,0 +1,21 @@
+package io.quarkus.oidc;
+
+import io.vertx.ext.web.RoutingContext;
+
+/**
+ * OIDC redirect filter which can be used to customize redirect requests to OIDC authorization and logout endpoints
+ * as well as local redirects to OIDC tenant error, session expired and other pages.
+ */
+public interface OidcRedirectFilter {
+    /**
+     * Filter OIDC redirect.
+     *
+     * @param routingContext the routing context that can be used to set additional cookies.
+     * @param oidcConfig current OIDC tenant configuration.
+     * @param redirectUri the redirect uri which filters may update by adding additional query parameters.
+     * @param tokenClaims the decoded JWT token claims in JSON format. If necessary, implementations can convert it to JSON
+     *        object.
+     * @return redirectUri the redirect uri which can have additional query parameters added.
+     */
+    String filter(RoutingContext routingContext, OidcTenantConfig oidcConfig, String redirectUri);
```

Comment:
It's not really clear to me why we should be returning a String here. What happens if you have more than a filter in your application? Last filter wins? How do you ensure order?

Reply:
Hi @gastaldi 
There is no ordering, the reason this filter returns `String` is to let filters add extra query parameters, which are unordered and the next filter which runs will see the original `redirectUri` or the one modified by the filters which run before it.

By the way, instead of returning String I've been considering returning `Map<String, String>` for those filters which want to add query parameters just return a Map, but then it is not clear if the next filter should see the extra query parameters as part of their `redirectUri` input, or all filters should see the very original `redirectUri`, so I just settled on the current approach for now, as described above...

Does it clarify it ?  Would you prefer returning a Map of query parameters ?

Reply:
@gastaldi Let me add one more filter to add another query parameter, and show how the test is impacted

Reply:
@gastaldi I've updated the test to have 2 filters now, so while at the assertion end the variation can happen, on the actual target end (see `TenantRefresh#sessionExpired`) it does not matter at all, query ordering can not be relied upon in general

Reply:
Sorry if it is not what you meant

Reply:
I think a `Map` would make more sense. IMHO it sucks to concat query param strings, especially when they need to be encoded, etc.

Reply:
But I think it needs to be a `MultiMap`, no?

Reply:
Yeah, this is what I've been hesitating about. But returning a Map poses a dilemma,  what if the first filter returns `Map.of("a", "a1")` but the next one returns `Map.of("a", "a2")`. We can set the rules that in this case we will have 2 values for `a`. But the problem is this 2nd filter does not know if `a=a1` already exists unless we add it to redirectUri but then we'd need to keep reparsing it...
So I wonder,  shout we have a void response but an extra MultiMap input which will be visible to all filters ?
The redirect uri will be updated with this multi map once all the filters are run ?

Reply:
Let's take one step back: Maybe I am missing something, but having a filter whose purpose is to add new query parameters seems a bit limited API-wise. I see in your examples that you can create cookies, header parameters, etc. I was expecting a similar API to include new Query/path parameters (eg. Users may ask: why do I need to invoke `RoutingContext.response().createCookie(...)` but for query parameters is different?).

Also, speaking about additional headers, https://quarkus.io/guides/http-reference#additional-http-headers may be the answer here

Reply:
@gastaldi Let me clarify, An OIDC centric interface is required, main reason, it must work in a multi-tenant OIDC setup, generic Vert.x filters, routes, or exception mappers don't work unfortunately. These filters should be able to make decisions on how to customize things given the current OIDC tenant configuration. 
Well, we may ask users, get the current OIDC config from RoutingContext in your custom Vert.x filter, which is not great and such filters will run for all responses, redirects, normal responses. That said, it probably won't work anyway, as Quarkus Security ends the response for several types of redirects.

With the OIDC centric solution, we can do some nice optimizations without users having to worry about making their Vert.x filter called in the right order. For example, right now there are about 5 possible redirects in the OIDC in various situations (1 - OIDC auithorization endpoint, 2 - OIDC logout endpoint, 3 - Local redirect to drop OIDC parameters after the user is returned back from OIDC auithorization 4 -  Local redirect to the error page of the user is returned back from OIDC auithorization with an error, 5 - which is main reason behind this PR - local redirect to the session expired page) - so as you can see from the test source, the test filter checks if this redirect matches a given request and only then do some work. But we let users provide filters for specific redirects only:
```
@SessionExpiredPage
public class SessionExpiredRedirectFilter implements OidcRedirectFilter {}
```
without having to be aware of the actual session expired page path and write `if (redirectUri.endsWith(somePath))`...

As far as query parameters, this is probably the least interesting case, in fact :-), I've seen users trying to intercept redirects with JAX-RS filters/mappers and set a response header, and we have a concrete practical use case for setting a cookie. But yes, the only other thing users may want to set on the redirect URI is, possibly, extra query parameters - note we already allow to configure such extra query parameters for the redirects 1. and 2 (OIDC authorization and logout redirects) - so here we are talking about dynamic query parameters for yet to be identified use cases :-)

So, `RoutingContext` gives us a mechanism to add headers or cookies, which does not impact the redirect URI itself. So an input `MultiMap` gives this tool. Yes, true, there is a disconnect, users can just add queries to this map, yet use low level RoutingContext to add headers  - well, we can add another Map, but what about cookies, which API would we use to represent cookies... - it is not worth it IMHO, `RoutingContext` already features in nearly every other OIDC interface because it can not be injected in most cases when these interfaces run.

What do you think ? Cheers

Reply:
Thanks for the explanation @sberyozkin. I was expecting a single API to apply all these changes in a filter, which would be `RoutingContext` for most of the things.
 
What if we introduced a `OIDCRoutingContext` as the single parameter to the filter method containing all these data, encapsulating the RoutingContext and introducing methods that we expect users will call without giving them the burden of having to learn the Vert.x internals? 

I apologize for insisting on this, but I believe it would be beneficial in the end

Reply:
Hey @gastaldi You are very welcome to insist on some specific design for this interface, np at all, it is a healthy discussion, thanks.

I'm hesitating though, we have `OidcRequestFilter` accepting Vert.x specific things like HttpRequest, we have `TokenStateManager`, `TokenInstrospectionCache` and `UserInfoCache` (and likely more interfaces) accepting `RoutingContext`, so introducing `OidcRoutingContext` does not feel right to me, `RoutingContext` there, but OIDC wrapper here. 
I guess if we had a Quarkus  Vert.x HTTP `RoutingContext` abstraction with only some key methods and generic common security interfaces like `HttpAuthenticationMechanism`, `IdentityProvider` had something like `SecurityRoutingContext` then it would've been different.

With `OidcRoutingContext`, how are we going to add cookies. Do we introduce `OidcCookie` as a wrapper around Vert.x `ServerCookie` ? And what if some user says then, actually I need to check HTTP request properties like current request uri or headers to decide how to customize he redirect, then we'll end up having nearly a full copy of RoutingContext, which is feasible, but I doubt it is worth it...

Thanks

Reply:
Well, my idea is `OIDCRoutingContext` to have a `getRoutingContext()` and all the other parameters in the filter method so you don't need to change the filter method signature when a new parameter is needed.

Reply:
I suppose we can name it `OidcRedirectContext` ?

So that instead of, say

```
void filter(RoutingContext context, OidcTenantConfig config, String redirectUri, MultiMap extraQueryParams) {
   if (context.getRoutingContext().response().addHeader()) {
       context.response().addHeader("a", "b");
   }
   if ("tenant-a".equals(config.tenantId())) {
      extraQueryParams.add("a", "b");
   }
}
```

we'd have
```
void filter(OidcRedirectContext context) {
   if (context.getRedirectUri().endsWith("/a")) {
     context.getRoutingContext().response().addHeader("a", "b");
   }
   if ("tenant-a".equals(context.getOidcConfig().tenantId())) {
      context.getRoutingContext().getExtraQueryParams.add("a", "b");
    } 
  
}
```

?

I suppose we can do it, and the signature is simpler.  Do we want to introduce a wrapper like this one though, in case of redirects, there  could only be 3 extra things that users can set during the redirect: headers, cookies (or if they want, `Set-Cookie` header), and extra query params, I can't quite imagine what else we'd need to add ? For better or worse, all other security interfaces, most of generic and OIDC ones, do not wrap, each interface (AFAIK) has several parameters, and so far I don't recall us having to change signatures :-). There were only cases related Uni-fying some of them, may be SecurityIdentityAugmentor is the only exception...

Reply:
Sure, that's just an idea. If that is not supposed to change often I believe we can keep the original signature, up to you :)

Reply:
@gastaldi Sure, thinking more about it, just because we have N parameters in the existing interfaces, why don't we try another approach and see how it goes :-)

Reply:
@gastaldi Have a look please at the new version (I like it more now :-) ), and if there is something you'd like to change, let me know please (test has a simple URL encoding checked too)